File: rottingOranges.cpp
//time complexity:O(m*n)
//space complexity:O(m*n)
//executed on leetcode: yes
//approach:using bfs
//any issues faced? yes, but understood it

class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        if(grid.size()==0)
            return 0;
        int freshorange=0;
        int time=-1;
        vector<vector<int>>dirs{{-1,0},{1,0},{0,-1},{0,1}};
        queue<pair<int,int>>q;
        int nrows=grid.size();
        int ncols=grid[0].size();
        for(int i=0; i<nrows; i++)
        {
            for(int j=0; j<ncols; j++)
            {
                if(grid[i][j]==2)
                    q.push({i,j});
                else if(grid[i][j]==1)
                    freshorange++;
            }
        }
        if(freshorange==0)
            return 0;
        while(!q.empty())
        {
            int size=q.size();
            for(int i=0; i<size; i++)
            {
                pair<int,int>curr=q.front();
                q.pop();
                int row=curr.first;
                int col=curr.second;
                for(vector<int>dir:dirs)
                {
                    int x=row+dir[0];
                    int y=col+dir[1];
                    if(x>=0 && y>=0 && x<grid.size() && y<grid[0].size() && grid[x][y]==1)
                    {
                        grid[x][y]=2;
                        freshorange--;
                        q.push({x,y});
                        
                    }
                }
            }
            time++;
        }
        if(freshorange==0)
            return time;
        else
            return -1;
    }
};

File: rotting-oranges.js
// Time Complexity : O(M*N)
// Space Complexity : O(M*N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {number[][]} grid
 * @return {number}
 */
var orangesRotting = function(grid) {
    if (grid == null || grid.length == 0) return 0;
    
    let m = grid.length,
        n = grid[0].length,
        q = [],
        fresh = 0,
        time = 0,
        dirs = [[0,1],[0,-1],[1,0],[-1,0]];
    
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] == 1) fresh++;
            if (grid[i][j] == 2) q.push([i, j]);
        }
    }
    if (fresh == 0) return 0;
    while (q.length) {
        let size = q.length;
        for (let i = 0; i < size; i++) {
            let curr = q.shift();
            for (const dir of dirs) {
                let r = dir[0] + curr[0],
                    c = dir[1] + curr[1];
                if (r >= 0 && r < m && c >= 0 && c < n && grid[r][c] == 1) {
                    grid[r][c] = 2;
                    fresh--;
                    q.push([r, c]);
                }
            }
        }
        time++;
    }
    if (fresh > 0) return -1;
    return time > 0 ? time - 1 : 0;
};
